Sources/WordPressUI/Extensions/UIImageView+Networking.swift

```diff
@@ -56,69 +55,73 @@ public extension UIImageView {
+    /// - parameters:
+    ///    - request: The request for the target image
+    ///    - placeholderImage: Image to be displayed while the actual asset gets downloaded.
+    ///    - success: Closure to be executed on success.
+    ///    - failure: Closure to be executed upon failure.
     ///
-    @objc func downloadImage(usingRequest request: URLRequest, placeholderImage: UIImage? = nil, success: ((UIImage) -> ())? = nil, failure: ((Error?) -> ())? = nil) {
-        cancelImageDownload()
-
-        let handleSuccess = { [weak self] (image: UIImage, url: URL) in
+    ///  - note: The callbacks are executed on the main queue. If you cancel the
+    ///  request, the previous callbacks are guaranteed not to be called: make
+    ///  sure to perform any cleanup before calling this method.
+    @objc func downloadImage(
+        usingRequest request: URLRequest,
+        placeholderImage: UIImage? = nil,
+        success: ((UIImage) -> ())? = nil,
+        failure: ((Error?) -> ())? = nil
+    ) {
+        assert(Thread.isMainThread, "The downloadImage method has to be called from the main thread")
+
+        let context = self.context
+        context.cancel()
+
+        let handleSuccess = { [weak self] (image: UIImage) in
             self?.image = image
             success?(image)
         }
 
         guard let url = request.url else {
-            if let placeholderImage = placeholderImage {
-                image = placeholderImage
+            if let placeholderImage {
+                self.image = placeholderImage
             }
-
             failure?(ImageDownloadError.noURLSpecifiedInRequest)
             return
         }
 
-        if let cachedImage = ImageCache.shared.getImage(forKey: url.absoluteString) {
-            handleSuccess(cachedImage, url)
+        let cacheKey = url.absoluteString
+        if let cachedImage = ImageCache.shared.getImage(forKey: cacheKey) {
+            handleSuccess(cachedImage)
             return
         }
 
         // Using the placeholder only makes sense if we know we're going to download an image
         // that's not immediately available to us.
-        if let placeholderImage = placeholderImage {
+        if let placeholderImage {
             self.image = placeholderImage
         }
 
-        let task = URLSession.shared.dataTask(with: request, completionHandler: { [weak self] data, response, error in
-            guard let data = data, let image = UIImage(data: data, scale: UIScreen.main.scale) else {
-                DispatchQueue.main.async {
-                    failure?(error)
-                    self?.downloadTask = nil
-                }
-                return
-            }
-
+        let taskId = context.getNextTaskId()
+        let task = URLSession.shared.dataTask(with: request) { data, response, error in
+            let image = data.flatMap(makeImage)
             DispatchQueue.main.async {
-                if response?.url == url {
```

Comment:
I'm not sure why it would even check this. If the server returns an incorrect image, it's an issue on the server. But the server could also redirect the request, and then it would fail for no good reason.

Reply:
I'm not sure either, I checked the last commit, https://github.com/wordpress-mobile/WordPressUI-iOS/commit/0726b0c19aa537a53a99ea5ec427716b2980871c, but I'm not sure what to make of it.